lib/TPP/Transforms/VectorContractToOuterproduct.cpp

```diff
@@ -0,0 +1,158 @@
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 2] !=
+            vector::IteratorType::parallel ||
+        iteratorTypes[iteratorTypes.size() - 1] !=
+            vector::IteratorType::reduction)
+      return rewriter.notifyMatchFailure(contractOp, "Not a gemm");
+
+    Location loc = contractOp.getLoc();
+    Value lhs = contractOp.getLhs();
+    Value rhs = contractOp.getRhs();
+    Value acc = contractOp.getAcc();
+
+    // Find the original tensor operands
+    auto lhsDefiningOp = lhs.getDefiningOp<vector::TransferReadOp>();
+    auto rhsDefiningOp = rhs.getDefiningOp<vector::TransferReadOp>();
+    if (!lhsDefiningOp || !rhsDefiningOp)
+      return failure();
+
+    Value lhsTensor = lhsDefiningOp.getSource();
+    Value rhsTensor = rhsDefiningOp.getSource();
+
+    auto lhsType = cast<ShapedType>(lhsTensor.getType());
+    auto rhsType = cast<ShapedType>(rhsTensor.getType());
+    auto accType = cast<VectorType>(acc.getType());
+
+    // Handle 3D subviews
+    int64_t M = accType.getDimSize(0);
+    int64_t N = accType.getDimSize(1);
+    int64_t K = lhsType.getDimSize(lhsType.getRank() - 1);
+
+    // Create constants
+    Value c0 = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+    Value c1 = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+    Value cK = rewriter.create<arith::ConstantIndexOp>(loc, K);
+
+    auto elementType = lhsType.getElementType();
+    FloatType floatType = cast<FloatType>(elementType);
+    Value f0 = rewriter.create<arith::ConstantFloatOp>(
+        loc, APFloat::getZero(floatType.getFloatSemantics()), floatType);
+
+    // Create the outer scf.for loop
+    auto forOp = rewriter.create<scf::ForOp>(
+        loc, c0, cK, c1, ValueRange{acc},
+        [&](OpBuilder &nestedBuilder, Location nestedLoc, Value iv,
+            ValueRange iterArgs) {
+          // Prepare indices for 3D subviews
+          SmallVector<Value, 3> lhsIndices, rhsIndices;
+          for (int i = 0; i < lhsType.getRank() - 2; ++i)
+            lhsIndices.push_back(c0);
+          lhsIndices.push_back(c0);
+          lhsIndices.push_back(iv);
+
+          for (int i = 0; i < rhsType.getRank() - 2; ++i)
+            rhsIndices.push_back(c0);
+          rhsIndices.push_back(iv);
+          rhsIndices.push_back(c0);
+
+          // Read vector slices using TransferReadOp
+          auto lhsSlice = nestedBuilder.create<vector::TransferReadOp>(
+              nestedLoc, VectorType::get({M}, lhsType.getElementType()),
+              lhsTensor, lhsIndices, f0);
+
+          auto rhsSlice = nestedBuilder.create<vector::TransferReadOp>(
+              nestedLoc, VectorType::get({N}, rhsType.getElementType()),
+              rhsTensor, rhsIndices, f0);
+
+          // Perform outer product
+          auto outerProduct = nestedBuilder.create<vector::OuterProductOp>(
+              nestedLoc, accType, lhsSlice, rhsSlice, iterArgs[0],
+              vector::CombiningKind::ADD);
```

Comment:
Combining kind should be propagated from the original `vector.contract` or lowering restricted to only match with `add`.

Reply:
For the moment only need ADD, will add an assert for it.

Reply:
Alright, just return failure on contract without `kind<add>`